### ISSUE TITLE
[DISCUSS] Specific event types

### ISSUE THREAD

**Issue by jhalterman** [jhalterman/failsafe#162]
Ahead of 2.0, I wanted to revisit the current event APIs. One problem is there are as many [as 9 methods](https://github.com/jhalterman/failsafe/blob/101140d7301969eeeb2d7a3791a043694febc019/src/main/java/net/jodah/failsafe/FailsafeConfig.java#L56-L135) to support a single event, with numerous variations with and without context, sync, async with and without executor, etc. We could drastically simplify things by having an event listener signature that accepts an event type:

```java
failsafe
  .with(retryPolicy)
  .onSuccess(event -> log.info("Connected to {}", event.result))
  .get(this::connect);
```

Events could be typed so that event listener statements don't need to declare the result type. An example event signature:

```java
/**
* @param <R> result type
*/
SuccessEvent<R> {
  public final R result;
  public final Throwable failure;
  public final int executions;
}
```

Thoughts?

**Comment by Tembrel** [comment]
First, is there any reason to prefer public final fields to accessor methods?

More importantly, what would an event type like `SuccessEvent` offer that `ExecutionContext` doesn't beyond the result type parameter and the result and failure objects? In other words, couldn't a revamped `ExecutionContext` *become* the event object?

(One would have to ensure that this `ExecutionContext`-like object is immutable, so that an asynchronous handler gets the context, result, and failure values that held when the event was fired.)

Specifically, if you're willing to give up the ability to write
```java
    .onSuccess(result -> log.info("Connected to {}", result))
```
and settle for
```java
    .onSuccess(event -> log.info("Connected to {}", event.getResult()))
```
then why not just add a type parameter to whatever `ExecutionContext` becomes after unifying sync and async stuff in #160, and promote `getLastResult` and `getLastFailure` to this type (since both `Execution` and `AsyncExecution` declare these methods identically)?

Then you could write something like this:
```java
    .onSuccess(ctx -> log.info("Connected to {}", ctx.getLastResult())
```
and you'd bring the 9 methods per event down to 3.

Furthermore, those 3 variants all have to do with how the handler is executed:
1. synchronously,
1. asynchronously using the scheduler configured for tries/retries, or
1. asynchronously using a different, supplied executor.

While I can see that one would want to have potentially different execution behavior for retries vs. event handlers, isn't it unlikely that one would want to vary the execution behavior __between__ event handlers? What about configuring a single scheduler/executor for event handlers, potentially different from that used for tries and retries (but not required to be), and then there could be one method per event: one `onSuccess`, one `onAbort`, etc. The default could be to use the same execution behavior as the tries/retries.

**Comment by jhalterman** [comment]
> First, is there any reason to prefer public final fields to accessor methods?

I've been programming too much Scala lately I suppose :)

But no, not really. I'd like to think that if/when Java gets value classes, then using public final fields won't seem so awkward if accessors really aren't needed. It does give me a little pause though.

> what would an event type like SuccessEvent offer that ExecutionContext doesn't beyond the result type parameter and the result and failure objects?

Good question. I suspect it might not offer much, if anything, beyond what ExecutionContext offers. It's possible that different events could expose different things, such as executionAttempts in retry related events and not in other events such as for CircuitBreakers, but I don't think there's much distinction.

> In other words, couldn't a revamped ExecutionContext become the event object?

I was thinking yes as well.

> then why not just add a type parameter to whatever ExecutionContext becomes after unifying sync and async stuff in #160, and promote getLastResult and getLastFailure to this type (since both Execution and AsyncExecution declare these methods identically)?

👍 I'd perhaps call it `.getResult`, but good observation that we can and probably should combine these things.

> While I can see that one would want to have potentially different execution behavior for retries vs. event handlers, isn't it unlikely that one would want to vary the execution behavior between event handlers? 

By different execution behavior do you mean using different executors? Maybe you can elaborate on this point.

> What about configuring a single scheduler/executor for event handlers, potentially different from that used for tries and retries (but not required to be), and then there could be one method per event: one onSuccess, one onAbort, etc. The default could be to use the same execution behavior as the tries/retries.

The 3 variants you mentioned are indeed the 3 that would make us fully featured if we go forward with event types. Personally, I started to question the value of async event handlers at all. In a way, if an execution is async then the event handling that is done post-execution is async in that it would be performed on the same thread as the execution. We could kick off a separate thread (in whatever executor) for this, I'm just not sure of when that would be necessary/beneficial. Any thoughts?

**Comment by Tembrel** [comment]
> By different execution behavior do you mean using different executors? Maybe you can elaborate on this point.

I just meant "execution policy", specifically, the choice of thread to run the event handler on.


> [...] I started to question the value of async event handlers at all. For sync or async user submitted executions, we could simply perform event handling in the same thread after the user submitted execution completes. We could run it separately in some other executor (the same one used for executions/retries or something supplied just for events), I'm just not sure of when that would be necessary/beneficial.

For synchronous user-submitted executions, where the calling thread is blocked anyway, I agree completely. Anyone who is doing substantial work in their event handlers can offload to a different executor manually, e.g., with `CompletableFuture`.


> My suspicion is that for most use cases and most users, executing events on the common pool, with or without tagging them onto the end of a user-submitted execution (we could pick one behavior or the other), would be just fine. Any thoughts?

I guess using the same thread as the user-submitted asynch execution ("tacking on the end") allows the user to do the same thing as above, e.g.,

```java
    .onSuccess(ctx -> CompletableFuture.supplyAsync(() -> expensiveLogging(ctx)))
```
So if the user doesn't want to run event handlers in the same pool used for user-submissions, they would have this out in both cases.

Bottom line: Sounds good to me.

> I should mention, this topic came up in a separate issue and slightly different context, when discussing whether to use the same executor that is used for a user-submitted callable for executing policies asynchronously, such as a fallback function. See [#145 (comment)](https://github.com/jhalterman/failsafe/issues/145#issuecomment-451712206). My sense there was that it's probably not necessary to have a separate async execution for a policy, and that it can just follow on the end of the user submitted callable's execution.

Haven't had time to look at #145, but will.

**Comment by jhalterman** [comment]
> I guess using the same thread as the user-submitted asynch execution ("tacking on the end") allows the user to do the same thing as above, e.g.,

Definitely 👍

**Comment by jhalterman** [comment]
This is now in master.

I ended up not unifying FailsafeEvent and ExecutionContext (for now) since the semantics are usually different. For most events that fire right after an execution, it makes sense to call `event.getResult()` or `event.getFailure()`. `getLastResult()` makes more sense for ExecutionContext (which we can add to that type in the future), or for something that happens just prior to some execution (which onRetry does, but, I'll lean towards simplicity for all the other events).

**Comment by jhalterman** [comment]
Closing for now. Feel free to leave a comment and/or re-open if the current approach described [above](https://github.com/jhalterman/failsafe/issues/162#issuecomment-454660335) is problematic.
